ZKUI-53 & ZKUI-54: Add legal hold in the object properties

ChengYanJin opened the pull request:
ZKUI-53: Add balance-scale in the object list
ZKUI-54: Add legal hold toggle in the object properties

![image](https://user-images.githubusercontent.com/18453133/147644778-6d6fd177-f862-4b88-a254-dc15c7d87131.png)
![image](https://user-images.githubusercontent.com/18453133/147644833-ac0aaefb-6d9a-44c9-9699-f28cafcf94ed.png)

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/zenko-ui/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZKUI-54 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.1.2`


Please check the `Fix Version/s` of ZKUI-54, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

JBWatenbergScality commented on src/react/actions/s3object.js:
```diff
@@ -533,3 +543,23 @@ export function putObjectTagging(
+    dispatch(networkStart('Getting object legal hold'));
+    return zenkoClient
+      .putObjectLegalHold(bucketName, objectKey, versionId, isLegalHold)
+      .then(() => dispatch(getObjectMetadata(bucketName, objectKey, versionId)))
+      .then(() => dispatch(listObjects(bucketName, prefixWithSlash)))
```
Maybe we could just dispatch the action that select the object here temporarily while the url rework is not done ?

ChengYanJin replied:
that's right!!! I will just reverse L.559 and L.560.  🤩  
and problem solved!

JBWatenbergScality commented on src/react/databrowser/objects/details/Properties.jsx:
```diff
@@ -17,6 +19,10 @@ const TruncatedValue = styled(T.Value)`
 function Properties({ objectMetadata }: Props) {
+  const dispatch = useDispatch();
+  const prefixWithSlash = usePrefixWithSlash();
+  const isLegalHoldEnabled = objectMetadata.legalHoldStatus === 'ON';
```
I'm wondering wether we should not keep this ON logic on the client side and everywhere else use a boolean.
In src/js/S3Client.js we could change :
```
getObjectLegalHold(bucketName, objectKey, versionId) {
    const params = {
      Bucket: bucketName,
      Key: objectKey,
      VersionId: versionId,
    };

    return new Promise((resolve, reject) => {
      this.client.getObjectLegalHold(params, (error, data) => {
        if (error) {
          if (
            error.code === 'NoSuchObjectLockConfiguration' ||
            error.code === 'InvalidRequest'
          ) {
            return resolve('OFF');
          }
          return reject(error);
        }
        if (data.Status) {
          return resolve(data.Status);
        }
        return resolve('OFF');
      });
    });
  }
```

To 
```
getObjectLegalHold(bucketName, objectKey, versionId) {
    const params = {
      Bucket: bucketName,
      Key: objectKey,
      VersionId: versionId,
    };

    return new Promise((resolve, reject) => {
      this.client.getObjectLegalHold(params, (error, data) => {
        if (error) {
          if (
            error.code === 'NoSuchObjectLockConfiguration' ||
            error.code === 'InvalidRequest'
          ) {
            return resolve(false);
          }
          return reject(error);
        }
        if (data.Status) {
          return resolve(data.Status === 'ON');
        }
        return resolve(false);
      });
    });
  }
```

Then everywhere else we wouldn't have to take care of the boolean translation of the value

ChengYanJin replied:
Yeah. It makes sense to store the boolean `isLegalHoldEnabled`

JBWatenbergScality commented on src/react/databrowser/objects/details/Properties.jsx:
```diff
@@ -61,14 +67,53 @@ function Properties({ objectMetadata }: Props) {
+                  Locked ({objectMetadata.objectRetention.mode.toLowerCase()})
+                  <br />
+                  until {objectMetadata.objectRetention.retainUntilDate}
+                </>
```
I think it is missing the locked icon here for the discoverability of the icon meaning (which was the starting point of this refactoring

JBWatenbergScality replied:
I've seen it afterwards in commit https://github.com/scality/zenko-ui/pull/214/commits/540df519b27ffa9a9e66cf2d34a0a85fdb9afd9b

JBWatenbergScality commented on src/react/databrowser/objects/details/Properties.jsx:
```diff
@@ -61,14 +67,53 @@ function Properties({ objectMetadata }: Props) {
+                </>
+              )}
+              {objectMetadata.lockStatus === 'RELEASED' &&
+                `Released - since ${objectMetadata.objectRetention.retainUntilDate}`}
```
I think the unlocked icon is also missing here

JBWatenbergScality replied:
I've seen it afterwards in commit https://github.com/scality/zenko-ui/pull/214/commits/540df519b27ffa9a9e66cf2d34a0a85fdb9afd9b

JBWatenbergScality commented on src/react/databrowser/objects/details/Properties.jsx:
```diff
@@ -61,14 +67,53 @@ function Properties({ objectMetadata }: Props) {
+                {isLegalHoldEnabled && (
+                  <i
+                    className="fas fa-balance-scale"
+                    style={{ paddingLeft: '5px' }}
```
shouldn't we use 0.5rem instead ? Maybe zenko is not yet ready for the rem unit conversion, I'm just wonering

ChengYanJin replied:
I actually changed in the later commit with `spacing`
```
const Icon = styled.i`
  margin-left: ${spacing.sp4};
`;
```

JBWatenbergScality commented on src/react/databrowser/objects/details/Properties.jsx:
```diff
@@ -1,5 +1,5 @@
 // @noflow
-import Table, * as T from '../../../ui-elements/TableKeyValue';
+import Table, * as T from '../../../ui-elements/TableKeyValue2';
```
Is TableKeyValue still used after this change ? If not I think we could drop it and rename the TableKeyValue2 to TableKeyValue 🚀

ChengYanJin replied:
Sadly this TableKeyValue is still used by many other components :(

JBWatenbergScality replied:
Would be for next time then !

ChengYanJin committed:
ZKUI-54: Add get/putObjectLegalHold() in S3Client

ChengYanJin committed:
ZKUI-54: Dispatch actions to fetch legal hold for objects and metadata

ChengYanJin committed:
ZKUI-53: Add balance-scale icon in the object list

ChengYanJin committed:
ZKUI-53: Extract usePrefixWithSlash hook

ChengYanJin committed:
ZKUI-54: Add legal hold toggle in the Properties

Update the Properties list with TableKeyValue2 since we need the group
value `Information` and `Data protection`

Add the icon next to the Lock status

ChengYanJin commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **approve**

nicolas2bert commented on src/react/databrowser/objects/details/Properties.jsx:
```diff
@@ -1,74 +1,141 @@
+  max-width: 18rem;
+`;
+
+const Icon = styled.i`
```
NIT: I know ui-elements/Table as an `Icon` component you might want to reuse. Or you can introduce one in `TableKeyValue2`

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/1.1`



The following branches will **NOT** be impacted:

* `development/1.0`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.1`



The following branches have **NOT** changed:

* `development/1.0`



Please check the status of the associated issue ZKUI-54.

Goodbye chengyanjin.
